### PR TITLE
Triggered sample

### PR DESCRIPTION
**Pull request by ehfeng**
Triggered platform-specific sample errors, to improve on-boarding experience.

[Screencast of new flow](https://www.dropbox.com/s/v35rmkgz4bdn24k/triggered%20sample.mov?dl=0)

image

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/api/endpoints/project_create_sample.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from rest_framework.response import Response
+
+from sentry.api.bases.project import ProjectEndpoint, ProjectPermission
+from sentry.api.serializers import serialize
+from sentry.utils.samples import create_sample_event
+
+
+class RelaxedProjectPermission(ProjectPermission):
+    scope_map = {
+        'POST': ['project:write', 'project:admin'],
```
Comment (macqueen) [inline src/sentry/api/endpoints/project_create_sample.py]: it looks like this has the same permissions level as `ProjectPermission`, so why is this necessary?

File: src/sentry/api/endpoints/project_create_sample.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+
+class ProjectCreateSampleEndpoint(ProjectEndpoint):
+    permission_classes = [RelaxedProjectPermission]
```
Comment (macqueen) [inline src/sentry/api/endpoints/project_create_sample.py]: we use tuples for this pretty much everywhere else that i can find

File: src/sentry/api/endpoints/project_create_sample.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    def post(self, request, project):
+        has_project_write = (
```
Comment (macqueen) [inline src/sentry/api/endpoints/project_create_sample.py]: what is this checking that the permission class is not?

File: src/sentry/static/sentry/app/views/onboarding/utils.js [hunk 4]
```diff
@@ -99,7 +99,8 @@ const stepDescriptions = {
 };
 
 function getPlatformName(platform) {
-  return flattenedPlatforms.find(({id}) => platform == id).name;
+  let platform_data = flattenedPlatforms.find(({id}) => platform == id);
```
Comment (macqueen) [inline src/sentry/static/sentry/app/views/onboarding/utils.js]: camel case

File: src/sentry/static/sentry/app/views/stream.jsx [hunk 5]
```diff
@@ -623,17 +638,30 @@ const Stream = React.createClass({
...
+      : 'JavaScript';
     let sampleLink = null;
     if (this.state.groupIds.length > 0) {
       let sampleIssueId = this.state.groupIds[0];
+
       sampleLink = (
         <p>
           <Link to={`/${org.slug}/${project.slug}/issues/${sampleIssueId}/?sample`}>
-            {t('Or see a sample Javascript event')}
+            {t('Or see your sample %s event', platformName)}
```
Comment (macqueen) [inline src/sentry/static/sentry/app/views/stream.jsx]: i think you want `tct`

File: src/sentry/static/sentry/app/views/stream.jsx [hunk 6]
```diff
@@ -623,17 +638,30 @@ const Stream = React.createClass({
...
           </Link>
         </p>
       );
+    } else {
+      sampleLink = (
+        <p>
+          <a onClick={this.createSampleEvent.bind(this, project.platform)}>
+            {t('Create a sample %s event', platformName)}
```
Comment (macqueen) [inline src/sentry/static/sentry/app/views/stream.jsx]: here too

File: src/sentry/utils/samples.py [hunk 7]
```diff
@@ -100,14 +100,19 @@ def load_data(platform, default=None, timestamp=None, sample_name=None):
...
     #   * When a user clicks Test Configuration from notification plugin settings page,
     #     a fake event is generated to go through the pipeline.
-    sample_name = sample_name or platform
+    sample_name = sample_name or INTEGRATION_ID_TO_PLATFORM_DATA[platform]['name']
 
     data = None
-    for platform in (platform, default):
+    language = None
+
+    if INTEGRATION_ID_TO_PLATFORM_DATA.get(platform)['type'] != 'language':
```
Comment (macqueen) [inline src/sentry/utils/samples.py]: doing `.get` here does nothing because while you're avoiding a `KeyError`, you're just going to wind up with a `TypeError` from `['type']`

File: src/sentry/utils/samples.py [hunk 8]
```diff
@@ -100,14 +100,19 @@ def load_data(platform, default=None, timestamp=None, sample_name=None):
...
+        language = INTEGRATION_ID_TO_PLATFORM_DATA.get(platform).get('language')
```
Comment (macqueen) [inline src/sentry/utils/samples.py]: same problem here

**Review by macqueen — CHANGES_REQUESTED** [review]

File: src/sentry/api/endpoints/project_create_sample.py (hunk 1 above)
Comment (ehfeng) [inline src/sentry/api/endpoints/project_create_sample.py]: Good point.

File: src/sentry/utils/samples.py (hunk 7 above)
Comment (ehfeng) [inline src/sentry/utils/samples.py]: Fixed

File: src/sentry/utils/samples.py (hunk 8 above)
Comment (ehfeng) [inline src/sentry/utils/samples.py]: Fixed

File: src/sentry/static/sentry/app/views/stream.jsx (hunk 6 above)
Comment (ehfeng) [inline src/sentry/static/sentry/app/views/stream.jsx]: Fixed

File: src/sentry/static/sentry/app/views/stream.jsx (hunk 5 above)
Comment (ehfeng) [inline src/sentry/static/sentry/app/views/stream.jsx]: Fixed

File: src/sentry/static/sentry/app/views/onboarding/utils.js (hunk 4 above)
Comment (ehfeng) [inline src/sentry/static/sentry/app/views/onboarding/utils.js]: Fixed

File: src/sentry/api/endpoints/project_create_sample.py (hunk 3 above)
Comment (ehfeng) [inline src/sentry/api/endpoints/project_create_sample.py]: Removed

File: src/sentry/api/endpoints/project_create_sample.py (hunk 2 above)
Comment (ehfeng) [inline src/sentry/api/endpoints/project_create_sample.py]: Fixed

File: src/sentry/utils/samples.py [hunk 9]
```diff
@@ -100,14 +100,21 @@ def load_data(platform, default=None, timestamp=None, sample_name=None):
...
-    sample_name = sample_name or platform
+    sample_name = sample_name or INTEGRATION_ID_TO_PLATFORM_DATA[platform]['name']
 
     data = None
-    for platform in (platform, default):
+    language = None
+
+    if platform in INTEGRATION_ID_TO_PLATFORM_DATA and 'type' in INTEGRATION_ID_TO_PLATFORM_DATA[
+        platform
+    ] and INTEGRATION_ID_TO_PLATFORM_DATA[platform]['type'] != 'language':
```
Comment (macqueen) [inline src/sentry/utils/samples.py]: why not something like:

```
language = None
platform_data = INTEGRATION_ID_TO_PLATFORM_DATA.get(platform)
if platform_data is not None and platform_data['type'] != 'language':
    language = platform_data['language']

```

**Review by macqueen — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/views/stream.jsx [hunk 10]
```diff
@@ -623,17 +638,30 @@ const Stream = React.createClass({
...
+      : 'JavaScript';
     let sampleLink = null;
     if (this.state.groupIds.length > 0) {
       let sampleIssueId = this.state.groupIds[0];
+
       sampleLink = (
         <p>
           <Link to={`/${org.slug}/${project.slug}/issues/${sampleIssueId}/?sample`}>
-            {t('Or see a sample Javascript event')}
+            {tct('Or see your sample %s event', platformName)}
```
Comment (macqueen) [inline src/sentry/static/sentry/app/views/stream.jsx]: based on percy, i think you need to use `[variable]` for formatting with `tct`, also i might have been wrong and maybe `t` would've been fine. I thought `tct` was the one that supported interpolation, but it might just be better if you have multiple variables/are using react components

**Review by macqueen — COMMENTED** [review]
on that note, percy is showing a lot of changes. can you review and make sure they are all intentional?

File: src/sentry/utils/samples.py (hunk 9 above)
Comment (ehfeng) [inline src/sentry/utils/samples.py]: Fixed

File: src/sentry/api/endpoints/project_create_sample.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+class ProjectCreateSampleEndpoint(ProjectEndpoint):
+    permission_classes = (ProjectPermission, )
+
+    def post(self, request, project):
+        if project.platform:
+            event = create_sample_event(
+                project, platform=project.platform, default='javascript', level=0
+            )
+        else:
```
Comment (ehfeng) [inline src/sentry/api/endpoints/project_create_sample.py]: In the event the project.platform doesn't have a language that has a sample json, (C# or Elixir for example), I'd like it to default to JavaScript.

File: src/sentry/static/sentry/app/views/stream.jsx (hunk 10 above)
Comment (ehfeng) [inline src/sentry/static/sentry/app/views/stream.jsx]: I'm not sure what you mean here?

Also, I'm thinking of removing `platformName` entirely, as in the event it's a Rails project, we don't actually have a Rails sample event—only a Ruby sample event. Passing this information from the server `/samples/*.json` to the frontend is complicated and not of huge value.

File: src/sentry/api/endpoints/project_create_sample.py (hunk 11 above)
Comment (macqueen) [inline src/sentry/api/endpoints/project_create_sample.py]: ummm is that not what `default='javascript'` is doing?

Comment (ehfeng) [inline src/sentry/api/endpoints/project_create_sample.py]: Oh...right. Damn.

**Comment by ehfeng** [comment]
Yes, all Percy changes are intentional.

**Review by macqueen — APPROVED** [review]
